src/databricks/labs/ucx/hive_metastore/view_migrate.py

```diff
@@ -83,86 +83,104 @@ def __eq__(self, other):
+            except RecursionError as e:
+                logger.error(
+                    f"Cannot sequence views {views_to_migrate} given migration index {self._index}", exc_info=e
+                )
+                # By returning the current batches, we can migrate the views that are not causing the recursion
+                return batches
+            for view in next_batch:
+                views_sequenced.add(self._views[view])
+            batches.append(next_batch)
+            views_to_migrate.difference_update(next_batch)
         return batches
 
-    def _next_batch(self, views: set[ViewToMigrate]) -> set[ViewToMigrate]:
+    def _next_batch(
+        self, views: set[ViewToMigrate], *, views_from_previous_batches: set[TableView] | None
+    ) -> list[ViewToMigrate]:
+        """For sequencing algorithm see docstring of :meth:sequence_batches.
+
+        Raises:
+            RecursionError :
+                If an infinite loop is detected.
+        """
+        views_from_previous_batches = views_from_previous_batches or set()
         # we can't (slightly) optimize by checking len(views) == 0 or 1,
         # because we'd lose the opportunity to check the SQL
-        result: set[ViewToMigrate] = set()
+        result = []
         for view in views:
-            view_deps = set(view.dependencies)
-            self._check_circular_dependency(view, views)
-            if len(view_deps) == 0:
-                result.add(view)
-            else:
-                # does the view have at least one view dependency that is not yet processed ?
-                not_processed_yet = view_deps - self._result_tables_set
-                if len(not_processed_yet) == 0:
-                    result.add(view)
-                    continue
-                if not [
-                    table_view
-                    for table_view in not_processed_yet
-                    if not self._index.is_migrated(table_view.schema, table_view.name)
-                ]:
-                    result.add(view)
-        # prevent infinite loop
-        if len(result) == 0 and len(views) > 0:
-            raise ValueError(f"Invalid table references are preventing migration: {views}")
+            self._check_circular_dependency(view)
+            if len(view.dependencies) == 0:
+                result.append(view)
+                continue
+            # If all dependencies are already processed, we can add the view to the next batch
+            not_processed_yet = set(view.dependencies) - views_from_previous_batches
+            if len(not_processed_yet) == 0:
+                result.append(view)
+                continue
+            if all(self._index.is_migrated(table_view.schema, table_view.name) for table_view in not_processed_yet):
+                result.append(view)
+        if len(result) == 0 and len(views) > 0:  # prevent infinite loop
+            raise RecursionError(f"Unresolved dependencies prevent batch sequencing: {views}")
```

Comment:
ideally we'd not rely on exceptions for control flow, but let's continue with this one